core/src/test/java/org/elasticsearch/index/replication/ESIndexLevelReplicationTestCase.java

```diff
@@ -0,0 +1,478 @@
+import org.elasticsearch.test.ESTestCase;
+import org.elasticsearch.test.junit.annotations.TestLogging;
+import org.elasticsearch.threadpool.ThreadPool;
+import org.elasticsearch.transport.TransportResponse;
+import org.junit.After;
+import org.junit.Before;
+
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.Future;
+import java.util.concurrent.FutureTask;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.function.BiFunction;
+import java.util.function.Consumer;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+import static org.hamcrest.Matchers.empty;
+import static org.hamcrest.Matchers.equalTo;
+
+public class ESIndexLevelReplicationTestCase extends ESTestCase {
+
+    private ThreadPool threadPool;
+    final private Index index = new Index("test", "uuid");
+    final private ShardId shardId = new ShardId(index, 0);
+    final private Map<String, String> indexMapping = Collections.singletonMap("type", "{ \"type\": {} }");
+    protected final static RecoveryTargetService.RecoveryListener recoveryListener = new RecoveryTargetService.RecoveryListener() {
+        @Override
+        public void onRecoveryDone(RecoveryState state) {
+
+        }
+
+        @Override
+        public void onRecoveryFailure(RecoveryState state, RecoveryFailedException e, boolean sendShardFailure) {
+            fail(ExceptionsHelper.detailedMessage(e));
+        }
+    };
+
+
+    @TestLogging("index.shard:TRACE,index.replication:TRACE,indices.recovery:TRACE")
+    public void testIndexingDuringFileRecovery() throws Exception {
+        try (ReplicationGroup shards = createGroup(randomInt(1))) {
+            shards.startAll();
+            int docs = shards.indexDocs(randomInt(50));
+            shards.flush();
+            IndexShard replica = shards.addReplica();
+            final CountDownLatch recoveryBlocked = new CountDownLatch(1);
+            final CountDownLatch releaseRecovery = new CountDownLatch(1);
+            final Future<Void> recoveryFuture = shards.asyncRecoverReplica(replica, new BiFunction<IndexShard, DiscoveryNode, RecoveryTarget>() {
+                @Override
+                public RecoveryTarget apply(IndexShard indexShard, DiscoveryNode node) {
+                    return new RecoveryTarget(indexShard, node, recoveryListener) {
+                        @Override
+                        public void renameAllTempFiles() throws IOException {
+                            super.renameAllTempFiles();
+                            recoveryBlocked.countDown();
+                            try {
+                                releaseRecovery.await();
+                            } catch (InterruptedException e) {
```

Comment:
if you catch IR you have to reset the interrupt status otherwise we keep on blocking